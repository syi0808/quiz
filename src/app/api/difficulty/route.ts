import { NextRequest, NextResponse } from 'next/server';
import cheerio from 'cheerio';

export async function GET(request: NextRequest) {
  const res = await fetch('https://opentdb.com/api_config.php');
  const html = await res.text();
  const $ = cheerio.load(html);

  const categories: { label: string; value: string }[] = $('select[name=trivia_difficulty]:first')
    .children()
    .toArray()
    .reduce((categories, optionElement) => {
      categories.push({
        label: (optionElement.children[0] as { data: string }).data,
        value: optionElement.attribs.value,
      });

      return categories;
    }, [] as { label: string; value: string }[]);

  return NextResponse.json(categories);
}
