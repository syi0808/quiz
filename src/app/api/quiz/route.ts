import { sleep } from '@/shared/utils/async';
import toCamelCase from '@/shared/utils/object';
import { NextRequest, NextResponse } from 'next/server';

export async function GET(request: NextRequest) {
  const { searchParams } = new URL(request.url);

  if (!searchParams.has('amount')) {
    searchParams.set('amount', '10');
  }

  if (searchParams.get('category') === 'any') {
    searchParams.delete('category');
  }

  if (searchParams.get('difficulty') === 'any') {
    searchParams.delete('difficulty');
  }

  const getResult = async () => {
    const res = await fetch(`https://opentdb.com/api.php?${searchParams.toString()}`);
    return await res.json();
  };

  do {
    const result = await getResult();

    if (result.response_code === 0) {
      return NextResponse.json(result.results.map(toCamelCase));
    } else {
      await sleep(100);
    }
  } while (1);
}
