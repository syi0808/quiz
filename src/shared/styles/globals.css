@font-face {
  font-family: 'Qilkabold';
  src: url('/fonts/Qilkabold.otf') format('opentype');
  font-weight: normal;
  font-style: normal;
}

body {
  font-family: 'Qilkabold';
}
