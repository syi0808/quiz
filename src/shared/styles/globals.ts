import * as sx from '@stylexjs/stylex';

export const globalStyles = sx.create({
  responsiveContainer: {
    width: {
      default: null,
      '@media (min-width: 1241px)': '1200px',
      '@media (max-width: 1240px)': '100%',
    },
    padding: {
      default: null,
      '@media (max-width: 1240px)': '0 20px',
    },
  },
});
