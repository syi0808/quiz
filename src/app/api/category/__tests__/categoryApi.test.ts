import { describe, test, expect } from '@jest/globals';
import { GET } from '../route';
import { NextRequest } from 'next/server';

describe('Get Category API', () => {
  test('receive and parsing categories correctly from html', async () => {
    const response = await GET({
      url: 'http://localhost:3000/api/category',
    } as NextRequest);

    const json = await response.json();

    expect(json).toMatchSnapshot();
  });
});
