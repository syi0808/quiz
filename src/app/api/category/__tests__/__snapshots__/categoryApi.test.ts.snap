// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Category API receive and parsing categories correctly from html 1`] = `
[
  {
    "label": "Any Category",
    "value": "any",
  },
  {
    "label": "General Knowledge",
    "value": "9",
  },
  {
    "label": "Entertainment: Books",
    "value": "10",
  },
  {
    "label": "Entertainment: Film",
    "value": "11",
  },
  {
    "label": "Entertainment: Music",
    "value": "12",
  },
  {
    "label": "Entertainment: Musicals & Theatres",
    "value": "13",
  },
  {
    "label": "Entertainment: Television",
    "value": "14",
  },
  {
    "label": "Entertainment: Video Games",
    "value": "15",
  },
  {
    "label": "Entertainment: Board Games",
    "value": "16",
  },
  {
    "label": "Science & Nature",
    "value": "17",
  },
  {
    "label": "Science: Computers",
    "value": "18",
  },
  {
    "label": "Science: Mathematics",
    "value": "19",
  },
  {
    "label": "Mythology",
    "value": "20",
  },
  {
    "label": "Sports",
    "value": "21",
  },
  {
    "label": "Geography",
    "value": "22",
  },
  {
    "label": "History",
    "value": "23",
  },
  {
    "label": "Politics",
    "value": "24",
  },
  {
    "label": "Art",
    "value": "25",
  },
  {
    "label": "Celebrities",
    "value": "26",
  },
  {
    "label": "Animals",
    "value": "27",
  },
  {
    "label": "Vehicles",
    "value": "28",
  },
  {
    "label": "Entertainment: Comics",
    "value": "29",
  },
  {
    "label": "Science: Gadgets",
    "value": "30",
  },
  {
    "label": "Entertainment: Japanese Anime & Manga",
    "value": "31",
  },
  {
    "label": "Entertainment: Cartoon & Animations",
    "value": "32",
  },
]
`;
